dali/pipeline/operators/detection/box_encoder.cc

```diff
@@ -0,0 +1,159 @@
+    // as this object will be the best (highest IOU), for this default box
+    ious[bbox_idx * M + best_idx] = 2.;
+  }
+  return ious;
+}
+
+Anchor ToCenterWH(const Anchor a) {
+  return {
+    0.5f * (a.x + a.z),
+    0.5f * (a.y + a.w),
+    a.z - a.x,
+    a.w - a.y};
+}
+}  // namespace detail
+
+template <>
+void BoxEncoder<CPUBackend>::FindMatchingAnchors(
+  const float *ious, const int64 N, const BBox *bboxes,
+  const int *labels, BBox *out_boxes, int *out_labels) {
+  // For every anchor we are looking, if there is a match
+  for (int anchor_idx = 0; anchor_idx < M_; ++anchor_idx) {
+    int best_idx = -1;
+    float best_iou = -1.;
+
+    // Looking for bbox with best IOU with current anchor
+    for (int bbox_idx = 0; bbox_idx < N; ++bbox_idx) {
+      if (ious[bbox_idx * M_ + anchor_idx] >= best_iou) {
+        best_iou = ious[bbox_idx * M_ + anchor_idx];
+        best_idx = bbox_idx;
+      }
+    }
+
+    out_labels[anchor_idx] = 0;
+
+    // Filter matches by criteria
+    // We only report a match, when IOU > criteria
+    if (best_iou > criteria_) {
+      out_boxes[anchor_idx] = bboxes[best_idx];
+      out_labels[anchor_idx] = labels[best_idx];
+    }
+
+    // Change to x, y, w, h per canonical SSD implementation
+    out_boxes[anchor_idx] = detail::ToCenterWH(out_boxes[anchor_idx]);
+  }
+}
+
+template <>
+void BoxEncoder<CPUBackend>::RunImpl(SampleWorkspace *ws, const int idx) {
+  const auto &bboxes_input = ws->Input<CPUBackend>(0);
+  const auto &labels_input = ws->Input<CPUBackend>(1);
+
+  const auto N = bboxes_input.dim(0);
+
+  auto bboxes = reinterpret_cast<const BBox *>(bboxes_input.data<float>());
+  auto labels = labels_input.data<int>();
+  auto anchors = reinterpret_cast<const Anchor *>(anchors_.data<float>());
+
+  auto ious = detail::CalculateIous(anchors, bboxes, N, M_);
+
+  // Create output
+  auto bboxes_output = ws->Output<CPUBackend>(0);
+  bboxes_output->set_type(anchors_.type());
+  bboxes_output->ResizeLike(anchors_);
+  auto out_boxes = reinterpret_cast<BBox *>(bboxes_output->mutable_data<float>());
+
+  auto labels_output = ws->Output<CPUBackend>(1);
+  labels_output->set_type(labels_input.type());
+  labels_output->Resize({M_});
+  auto out_labels = labels_output->mutable_data<int>();
+
+  // Copy default boxes (anchors) to output
+  TypeInfo type = anchors_.type();
+  type.Copy<CPUBackend, CPUBackend>(bboxes_output->raw_mutable_data(), anchors_.raw_data(),
+                                    anchors_.size(), 0);
+
+  FindMatchingAnchors(ious.data(), N, bboxes, labels, out_boxes, out_labels);
+}
+
+DALI_REGISTER_OPERATOR(BoxEncoder, BoxEncoder<CPUBackend>, CPU);
+
+DALI_SCHEMA(BoxEncoder)
+    .DocStr(
+        R"code("Encodes input bounding boxes and labels using set of default boxes (anchors) passed 
+        during op construction. Follows algorithm described in https://arxiv.org/abs/1512.02325 and 
```

Comment:
Remove indentation inside doc string or you will get strange doc content.

Reply:
Fixed